File: src_qt/rules/ConditionPropertyNode.cpp
#include <QHashIterator>

#include "ConditionPropertyNode.h"
#include "../Brain.h"

namespace CnotiMind
{

	ConditionPropertyNode::ConditionPropertyNode( const QString& property, const QString& value, ConditionOperator op, Brain* brain, QObject* parent ) :
		ConditionNode( property, value, op, brain, parent )
	{
	}


	/*
		Test the value of the property using the conditional operator
	*/
	bool ConditionPropertyNode::isTrue()
	{
		QHashIterator<QString, QString> it( _brain->_properties );
		while( it.hasNext() )
		{
			it.next();
			const QString& k = it.key();
			const QString& v = it.value();

			// Test if found the property
			if( _key.compare( k, Qt::CaseInsensitive ) == 0 )
			{
				// Try to convert the value from the variable to number
				bool ok;
				qreal valueNumeric = v.toDouble(&ok);

				if( ok && _isValueNumeric ) // if both values are numbers
				{
					switch( _operator )
					{
						case ConditionOperatorBigger: return _valueNumeric > valueNumeric;
						case ConditionOperatorBiggerOrEqual: return _valueNumeric >= valueNumeric;
						case ConditionOperatorSmaller: return _valueNumeric < valueNumeric;
						case ConditionOperatorSmallerOrEqual: return _valueNumeric <= valueNumeric;
						case ConditionOperatorEqual: return _valueNumeric == valueNumeric;
						case ConditionOperatorDifferent: return _valueNumeric != valueNumeric;
					}
					return false;
				}
				else // If one of the value is not a number
				{
					switch( _operator ) // it's a string, just to this 2 operators
					{
						case ConditionOperatorEqual: return QString::compare( _value, v, Qt::CaseInsensitive ) == 0;
						case ConditionOperatorDifferent: return QString::compare( _value, v, Qt::CaseInsensitive ) != 0;
					}
					return false;
				}
			}
		}

		// No equal variables found
		return false;
	}


	QString ConditionPropertyNode::info( int depth ) const
	{
		QString info;

		info += space(depth) + "Condition type=Property";

		return info + RuleNode::info(depth);
	}

	ConditionPropertyNode* ConditionPropertyNode::fromXML(const QString &qName, const QXmlAttributes &atts, Brain *brain, QObject *parent)
	{
		if(qName.compare( "Condition", Qt::CaseInsensitive) == 0)
		{
			QString type = atts.value( "type" );
			if( type.compare("Property") == 0 )
			{
				QString property = atts.value( "property" );
				QString value = atts.value( "compareValue" );
				ConditionOperator op = translateConditionOperator( atts.value( "operator" ) );

				return new ConditionPropertyNode( property, value, op, brain, parent );

			}
		}
	}


}

File: src_ios/CnotiMind/rules/ConditionPropertyNode.m
//
//  ConditionPropertyNode.m
//  SingleEmotion
//
//  Created by Gonçalo Rodrigues on 8/31/11.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "ConditionPropertyNode.h"

#import "Brain.h"

@implementation ConditionPropertyNode

- (id)init
{
    self = [super init];
    if (self) {
        // Initialization code here.
    }
    
    return self;
}


- (id) initWithPropertyAndValuAndOperatorAndBrainAndParent: (NSString*)aProperty 
                                                     value:(NSString*)aValue
                                                  operator:(enum ConditionOperator)aOperator
                                                     brain:(Brain*)aBrain 
                                                    parent:(id)aParent
{

    if (self = [super initWithKeyAndValueAndBrainAndParent:aProperty value:aValue operator:aOperator brain:aBrain parent:aParent]) {
    }

    return self;
}


- (BOOL) isTrue{

    NSEnumerator* eProperties = [[_brain properties] objectEnumerator];
    
    for (NSString* aKey in eProperties) {
        
        NSString* k = [NSString stringWithFormat:@"%@", aKey];
        NSString* v = [NSString stringWithFormat:@"%@", [[_brain properties] objectForKey:aKey]];
        
        // Test if found the property
        if (![_key caseInsensitiveCompare:k]) {
            // Try to convert the value from the variable to number
            
            BOOL ok;
            float valueNumeric;
            if ([self isNumeric:v]) {
                ok = true;
                valueNumeric = [v floatValue];
            }
            
            if (ok && _isValueNumeric) { // if both values are numbers
                switch( (int)_operator )
                {
                    case ConditionOperatorBigger: return _valueNumeric > valueNumeric;
                    case ConditionOperatorBiggerOrEqual: return _valueNumeric >= valueNumeric;
                    case ConditionOperatorSmaller: return _valueNumeric < valueNumeric;
                    case ConditionOperatorSmallerOrEqual: return _valueNumeric <= valueNumeric;
                    case ConditionOperatorEqual: return _valueNumeric == valueNumeric;
                    case ConditionOperatorDifferent: return _valueNumeric != valueNumeric;
                }
                return false;            
            }
            else { // If one of the value is not a number
                switch( (int)_operator ) // it's a string, just to this 2 operators
                {
                    case ConditionOperatorEqual: return ![_value caseInsensitiveCompare:v];
                    case ConditionOperatorDifferent: return [_value caseInsensitiveCompare:v];
                }
                return false;
            }
            
        }
    }
    
    // No equal variables found
    return false;
}


- (NSString*) info:(int)aDepth{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ Condition type=Property %@", space, [super info:aDepth]];
    
    return info;
}

@end
